src/txkube/test/test_authentication.py

```diff
@@ -68,6 +72,35 @@
 -----END CERTIFICATE-----
 """
 
+
+def createReactor():
+    """
+    Twisted 17.1.0 and higher requires a reactor which implements
+    ``IReactorPluggableNameResolver``.
+    """
+
+    class Resolution(object):
+        name = None
+
+    class _FakeResolver(object):
+        def resolveHostName(self, resolutionReceiver, hostName, *args,  **kwargs):
+            portNumber = kwargs.pop('portNumber')
+            r = Resolution()
+            r.name = hostName
+
+            resolutionReceiver.resolutionBegan(r)
+            resolutionReceiver.addressResolved(IPv4Address('TCP', 'example.invalid.', portNumber))
```

Comment:
I suppose that since there's now an explicit resolution step going on (whereas previously the hostname would be passed to `connectTCP`), the result here should be an IP address instead of a hostname?  The hostname and the IP address are both sloppily represented as native strings but they're pretty distinctly different things.

Pick an address from one of the `TEST-NET` ranges, probably: https://en.wikipedia.org/wiki/IPv4#Special-use_addresses.

Also, there should be some kind of assertion made about the value of `hostName` somewhere.  It is currently essentially ignored, as far as I can tell.  Previously, each test asserted that "example.invalid." was being connected to (a value taken from an environment variable).  Now no matter what hostname is passed in, it will be resolved to the hard-coded value passed to this `addressResolved` method.  So the coverage of the hostname being selected properly is being removed.

A well-behaved fake resolver, I suppose, would know about just the relevant fake hostnames and return a resolution error for any others (in whatever way is correct for the resolver interface now, I haven't checked, but I guess there's some other method).

Reply:
So `resolveHostname` https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IHostnameResolver.html#resolveHostName returns an `IResolutionReceiver`: 
https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IResolutionReceiver.html
not an IP address.

I'm new to all these interfaces, but let me see what I can do to clean this up.